Challenge repo init command

cacama-valvata opened the pull request:
Opening a premature PR on this so that I can get midway feedback and get most of my "TODO" questions answered this Sunday (not all of the TODOs are for y'all but most of them are, feel free to comment on any you want to give insight on). Please add your answers to the TODOs via comments in a review on this PR since I can't be there this Sunday

detjensrobert commented on src/commands/init.rs:
```diff
@@ -0,0 +1,337 @@
+    s3_secretaccesskey: String
+}
+
+pub fn run(_interactive: &bool, _blank: &bool) // TODO: is there a way to set options at mutually exclusive?
```
There is, `conflicts_with` in clap-derive ([docs](https://docs.rs/clap/latest/clap/_derive/_tutorial/chapter_3/index.html#argument-relations))

cacama-valvata replied:
done

detjensrobert commented on src/commands/init.rs:
```diff
@@ -0,0 +1,337 @@
+    let defaults_resources_cpu;
+    let defaults_resources_memory;
+    let mut points_difficulty = Vec::new();
+    let mut deploy_profiles = Vec::new();
```
This should probably be an `mut init_vars` struct instead of a bunch of individual variables, since these are all getting set into one after the prompts.

(ditto for points and profile sections)

cacama-valvata replied:
"done"

irl convo to perform the inquire calls within the struct initializer

detjensrobert commented on src/commands/init.rs:
```diff
@@ -0,0 +1,337 @@
+    //                 })
+    defaults_resources_memory = inquire::Text::new("Default limit of memory per challenge")
+                        .with_help_message("The maximum limit of memory resources per instance of challenge deployment (\"pod\").")
+                        .prompt()?;
```
I'm not sure if these need to be prompted for. Most challenges are going to be fine with a reasonable default resource limit, and I think we should figure out what that default is and always include it. I think users will very rarely need to change these here for all challenges, and that can be left to manual edits.

cacama-valvata replied:
provided a sane default and inquire placeholder, but still prompting for it in case

detjensrobert commented on src/commands/init.rs:
```diff
@@ -0,0 +1,337 @@
+                        .with_help_message("The maximum limit of memory resources per instance of challenge deployment (\"pod\").")
+                        .prompt()?;
+
+    println!("You can define several challenge difficulty classes below.");
```
Wrong header message?

cacama-valvata replied:
done (fixed)

detjensrobert commented on src/commands/init.rs:
```diff
@@ -0,0 +1,337 @@
+    }
+
+    // TODO -- does not compile because options does not yet implement serialize
+    // comment out if wanting to test the rest of the code
```
Adding `#[derive(Serialize)]` to the other structs fixes this

cacama-valvata replied:
done, and switched to minijinja instead of terra

detjensrobert commented on src/cli.rs:
```diff
@@ -69,4 +69,17 @@ pub enum Commands {
         registry: bool,
     },
+
+    /// Copy an initial rcds.yaml to the current working directory.
```
Since this command can do interactive prompts aside from non-interactive copy/write-out, this description should a bit more generic, something like

```suggestion
    /// Set up a new rcds.yaml config in the current working directory
```

cacama-valvata replied:
done

detjensrobert commented on src/cli.rs:
```diff
@@ -69,4 +69,17 @@ pub enum Commands {
+    ///
+    /// If interactive is enabled, then it will prompt for the various fields of the config file. If left disabled, then it will copy it out with fake data of the expected format.
+    /// 
+    /// If blank is enabled, then it will copy out the file without any fields set. If left disabled, it will write the default non-interactive example config to file.
```
These should be on the arguments since they are describing what they do.

cacama-valvata replied:
done

detjensrobert commented on src/commands/init.rs:
```diff
@@ -0,0 +1,337 @@
+
+struct init_vars
+{
+    flag_regex: String, // TODO: make all of these `str`s if it compiles
```
Structs need to own the data they contain, so these do all have to be `String`s.

cacama-valvata replied:
done

detjensrobert commented on src/commands/init.rs:
```diff
@@ -0,0 +1,337 @@
+    // TODO: would the cluster not use a token of some sort?
+    registry_cluster_pass = inquire::Password::new("Container registry password (CLUSTER'S):")
+                    .with_help_message("The cluster's password to the remote container registry, which it will use to pull containers from.")
+                    .prompt()?;
```
Kubernetes uses the same `base64("username:password")` credential format as Docker does, so we're matching that here.

cacama-valvata replied:
thanks for the insight, done

detjensrobert commented on src/commands/init.rs:
```diff
@@ -0,0 +1,337 @@
+
+    // TODO: do we actually want to be in charge of these credentials vs letting the container building utility take care of it?
+    registry_build_pass = inquire::Password::new("Container registry password (YOURS):")
+                    .with_help_message("Your password to the remote container registry, which you will use to push containers to.") // TODO: could this support username:pat too?
```
To me, "your" is a bit confusing since the config calls this build credentials -- "local/build" might work better?

cacama-valvata replied:
done

detjensrobert reviewed (commented):
Structure comments:
- So far, I've tried to keep the bulk of the logic for comands in their own folders, and `run()` orchestrates which thing to call based on the command line flags (i.e. `build::run()` calls build_image() and then push_images() if `--push` was given). I think the `*_init()` functions and the struct definitions should go in their own folder/file to follow this. Thoughts from the rest of the team?

cacama-valvata committed:
init command main workflow

cacama-valvata committed:
init command, formatted

cacama-valvata committed:
init command, lump external-dns configuration options into 'not our problem' style

cacama-valvata committed:
init cmd added inquire placeholders and cleaned up prompt and help messages

cacama-valvata committed:
init command - allowed difficulty class to be a string instead of a number identifier

cacama-valvata committed:
changed structs to upper pascal case

cacama-valvata committed:
cleaning up & addressing robert's feedback in the old branch/PR

detjensrobert committed:
Fix result handling in init command runner

Signed-off-by: Robert Detjens <github@detjens.dev>

detjensrobert committed:
Clippy fixes, use Default trait for blank

Signed-off-by: Robert Detjens <github@detjens.dev>

detjensrobert committed:
Tracing logs and use render_strict helper

Signed-off-by: Robert Detjens <github@detjens.dev>

detjensrobert committed:
use to_string, inquire default values

Signed-off-by: Robert Detjens <github@detjens.dev>

detjensrobert committed:
simplify init imports

Signed-off-by: Robert Detjens <github@detjens.dev>

detjensrobert committed:
Show external-dns config notice as warning

Signed-off-by: Robert Detjens <github@detjens.dev>

detjensrobert committed:
Use our internal config structs for init handling -- example init

We already have an existing struct used to parse the rcds.yaml config,
use that to build the example/interactive config instead of creating
an almost-identical struct.

So far this is only done for the example values / non-interactive init.

Signed-off-by: Robert Detjens <github@detjens.dev>

detjensrobert committed:
Use our internal config structs for init handling -- blank init

Signed-off-by: Robert Detjens <github@detjens.dev>

detjensrobert committed:
Convert points difficulty category name to a string

Having this be a string is more user friendly than a number, and this
branch was originally set up to do that.

Signed-off-by: Robert Detjens <github@detjens.dev>

detjensrobert committed:
Use our internal config structs for init handling -- interactive prompts

Signed-off-by: Robert Detjens <github@detjens.dev>

detjensrobert committed:
Update rcds config template with help comments and use new structs

Signed-off-by: Robert Detjens <github@detjens.dev>

detjensrobert committed:
Handle empty profile/points correctly

Signed-off-by: Robert Detjens <github@detjens.dev>

detjensrobert committed:
Add trailing template comment to preserve trailing newline when blank

Signed-off-by: Robert Detjens <github@detjens.dev>

detjensrobert committed:
Make default `init` behaviour interactive

Blank and placeholder template files should be set explicitly by the user.

Signed-off-by: Robert Detjens <github@detjens.dev>

detjensrobert committed:
Fix blank points and reuse default tag format from parsing

Signed-off-by: Robert Detjens <github@detjens.dev>

detjensrobert committed:
Switch to serde_yaml_ng

serde_yml turns out to be maintained weirdly (issues disabled, weird refactors)
and is now archived anyway. _ng is what the community is now recommending.

Signed-off-by: Robert Detjens <github@detjens.dev>

detjensrobert committed:
format interactive prompt structure, document rustfmt failings

Signed-off-by: Robert Detjens <github@detjens.dev>

detjensrobert committed:
Prompt user for tag format

Signed-off-by: Robert Detjens <github@detjens.dev>

detjensrobert committed:
reword formatting comment

Signed-off-by: Robert Detjens <github@detjens.dev>

detjensrobert committed:
Add reference link to docs site before prompting

Signed-off-by: Robert Detjens <github@detjens.dev>

detjensrobert committed:
Consistent punctuation on prompts

Signed-off-by: Robert Detjens <github@detjens.dev>

detjensrobert commented on src/init/mod.rs:
```diff
@@ -0,0 +1,290 @@
+
+    let difficulty_names; // set during `points` prompt later
+
+    // FORMATTING NOTE: Some of these help messages cause rustfmt to silently
+    // fail to format this struct definition. Commenting out the marked
+    // help_message lines temporarily will let the formatting work.
+    //
+    // Ref:
+    // - https://github.com/rust-lang/rustfmt/issues/6687
+    // - https://github.com/rust-lang/rustfmt/issues/3863
```
Rustfmt currently has some issues with formatting this file due to line length limitations. The issues linked here have more context (https://github.com/rust-lang/rustfmt/issues/6687, https://github.com/rust-lang/rustfmt/issues/3863) -- TLDR rustfmt silently fails on really long lines/method chains and thus fails to  format this file unless the lines marked with `// too long to format` are commented out.

@cacama-valvata and I discussed offline a while ago about this and he recommended that this is is documented in the Discord rather than a code comment here since this seems unprofessional. I disagree with that point and prefer to have it documented here so that it cannot get lost or forgotten and have to figure this all out again in the future, because this is a bug in rustfmt and not (personally) a code smell.

Do any of the rest of you have thoughts on if this should be documented and where?

Kajmany replied:
TIL this is a problem. I wonder how often it's happened to me and I never noticed before?

I tried `format_strings` with the nightly toolchain and it doesn't help here. I don't think it's worth breaking the strings manually, either. I think the status quo is fine even if the `too long to format` comments aren't  necessary.

Kajmany commented on src/init/mod.rs:
```diff
@@ -0,0 +1,290 @@
+            },
+            cluster: config::UserPass {
+                user: inquire::Text::new("Container registry 'cluster' user:")
+                    .with_help_message(
+                        "The username that the cluster will use to pull locally-built containers.",
+                    )
```
I would personally just use a regex to comment these all out at once so I could format. This is the only help message that's not contained on a single line. Not a big deal to change it manually and do that.

detjensrobert reviewed (commented):
Also a note: the placeholder/blank config structs could be done via the `Default` trait on the object, but only one of them can so I have not done that. Should one of them be the Default, and if so which?

Kajmany commented on src/init/mod.rs:
```diff
@@ -0,0 +1,290 @@
+            },
+        },
+
+        points: {
+            println!("You can define several challenge difficulty classes below.");
+            let mut again = inquire::Confirm::new("Do you want to provide a difficulty class?")
+                .with_default(false)
+                .prompt()?;
+
+            println!("Challenge points are dynamic. For a static challenge, simply set minimum and maximum points to the same value.");
+            let mut points = vec![];
+            while again {
+                let points_obj = config::ChallengePoints {
+                    difficulty: inquire::Text::new("Difficulty class:")
+                        .with_validator(inquire::required!("Please provide a name."))
+                        .with_help_message("The name of the difficulty class.")
+                        .with_placeholder(example_values::POINTS_DIFFICULTY)
+                        .prompt()?,
+                    min: inquire::CustomType::<i64>::new("Minimum points:")
+                        .with_error_message("Please type a valid number.") // default parser calls std::u64::from_str
+                        .with_help_message("The minimum number of points that challenges within this difficulty class are worth.") // too long to format
+                        .with_default(example_values::POINTS_MIN)
+                        .prompt()?,
+                    max: inquire::CustomType::<i64>::new("Maximum points:")
+                        .with_error_message("Please type a valid number.") // default parser calls std::u64::from_str
+                        .with_help_message("The maximum number of points that challenges within this difficulty class are worth.") // too long to format
+                        .with_default(example_values::POINTS_MAX)
+                        .prompt()?,
+                };
+                points.push(points_obj);
+
+                again = inquire::Confirm::new("Do you want to provide another difficulty class?")
+                    .with_default(false)
+                    .prompt()?;
+            }
+            // save owned copy of difficulty category names for use below
+            difficulty_names = points.iter().map(|p| p.difficulty.clone()).collect_vec();
+            points
+        },
```
These don't appear to be serialized right, and it's not obvious to me why. Here's some output I got from typing in three manually:

```
# The list of different difficulties that challenges can be assigned, and how
# many points challenges of that difficulty class should be worth. All
# challenges use dynamic scoring; for static points set both min and max to the
# same value.
points:
  - difficulty: "easy"
    min: 200
    max: 500- difficulty: "manual"
    min: 500
    max: 1000- difficulty: "ez"
    min: 999999999999
    max: -900
```
It's probably also ideal to validate the input against POINTS_MAX and POINTS_MIN, and maybe for positive-only points?

detjensrobert replied:
good catch, fixed the whitespace stripping in the template for that

detjensrobert committed:
Fix rendering of multiple difficulty modes in init template

Signed-off-by: Robert Detjens <github@detjens.dev>

detjensrobert committed:
add tests for blank/placeholder init templating

Signed-off-by: Robert Detjens <github@detjens.dev>

cacama-valvata commented on src/init/mod.rs:
```diff
@@ -0,0 +1,316 @@
+    }
+}
+
+pub fn example_init() -> config::RcdsConfig {
```
rename to placeholder_init to match the cli option

cacama-valvata commented on src/asset_files/rcds.yaml.j2:
```diff
@@ -0,0 +1,82 @@
+# Registry configuration for challenge images.
+registry:
+  domain: "{{ registry.domain }}"
+  # This is the default tag format; it will create a separate image for each
```
This seems extremely verbose for comments. Instead place in online documentation?

Also can we give this a default value when omitted? This seems that it should work for most people, so I don't think we should include this for the placeholder option.

detjensrobert committed:
clean up super verbose tag format comment

Signed-off-by: Robert Detjens <github@detjens.dev>

detjensrobert committed:
match placeholder function to cli flag name

Signed-off-by: Robert Detjens <github@detjens.dev>

detjensrobert committed:
Add force flag to overwrite any existing rcds config on init

Signed-off-by: Robert Detjens <github@detjens.dev>

detjensrobert reviewed (approved):
consensus from server, good to merge now and follow up with other prs for any changes
